[bgfx] [bimg] [bx] Add new port

BurningEnlightenment opened the pull request:
Add a port for [bgfx](https://github.com/bkaradzic/bgfx). As opposed to #9369 this _mostly_ doesn't use vendored packages, because I replaced the GENIe buildsystem with a straight forward cmake buildsystem with the added benefit that cmake import targets can be provided for everything.

I rewrote the cmake files from scratch and didn't use the ones provided by [JoshuaBrookover/bgfx.cmake](https://github.com/JoshuaBrookover/bgfx.cmake) in order to avoid potential licensing/CLA issues.

However, it would be absolutely awesome if @JoshuaBrookover could review the approach I've taken. Anyways, many thanks for maintaining `bgfx.cmake`!

***
The `bimg` and `bgfx` ports currently rely on some vendored packages which are currently not available in vcpkg. Also I needed to update some of the existing ports, obviously those updates need to be merged first:
* #10426
* #10427
* #10441
* #10442
* #10457


The port has been tested with the x64-windows-static and x64-linux triplets. OSX support is largely lacking due to not having anything to test with.
A "preview" branch with all changes merged can be found here: [bgfx-asm](https://github.com/BurningEnlightenment/vcpkg/tree/dev/Enlightenment/bgfx-asm)
***
Resolves #1303
Closes #9369

BurningEnlightenment committed:
[bx] Add bx port version af9ccf@2020-03-14

Build the amalgated version of the library using a custom CMake project
in order to provide cmake import configs.

BurningEnlightenment committed:
[bimg] Add bimg port version 2020-03-15

Build the library using a custom CMake project in order to provide cmake
import configs.

BurningEnlightenment committed:
[bgfx] Add initial port with 78675e@2020-03-15

Build the amalgated version of the library using a custom CMake project
in order to provide cmake import configs.

- Provide feature flags for the most common renderer backends.
- Provide feature flags for building the examples and/or tools.

dan-shaw commented:
/azp run

NancyLi1013 commented on ports/bgfx/portfile.cmake:
```diff
@@ -0,0 +1,82 @@
+file(COPY "${CURRENT_PORT_DIR}/CMakeLists.txt" DESTINATION "${SOURCE_PATH}")
+
+vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
+    FEATURES
```
This key work `FEATURES` is not needed to add. Please remove it.

NancyLi1013 commented on ports/bgfx/portfile.cmake:
```diff
@@ -0,0 +1,82 @@
+        d3d12    BGFX_WITH_D3D12
+        examples BGFX_WITH_EXAMPLES
+        tools    BGFX_WITH_TOOLING
+    INVERTED_FEATURES
```
Since this is no` INVERTED_FEATURES`, this key word is not needed either.

NancyLi1013 commented on ports/bgfx/portfile.cmake:
```diff
@@ -0,0 +1,82 @@
+#  * glsl-optimizer  -- https://github.com/aras-p/glsl-optimizer
+#  * iconfontheaders -- https://github.com/juliettef/IconFontCppHeaders
+#  * meshoptimizer   -- https://github.com/zeux/meshoptimizer
+#  * sdf             -- unknown origin
```
Can these comments be removed here?

BurningEnlightenment replied:
I know about the "no comments" guideline, however I'd rather let them stay, because they clearly identify potential port improvements and will help future maintainers implementing these.

NancyLi1013 commented on ports/bimg/CONTROL:
```diff
@@ -0,0 +1,10 @@
+
+Feature:
+Description:
+Build-Depends:
```
If there is no any feature to be added, please remove these lines.

NancyLi1013 commented on ports/bimg/portfile.cmake:
```diff
@@ -0,0 +1,49 @@
+
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO "bkaradzic/bimg"
```
There is no need to use double quotes in `REPO`.

NancyLi1013 commented on ports/bimg/portfile.cmake:
```diff
@@ -0,0 +1,49 @@
+#   * etc2          -- unknown source; defines symbols like `::g_table`, high likelyhood of ODR violations
+#   * iqa           -- https://sourceforge.net/projects/iqa or https://github.com/tjdistler/iqa
+#   * nvtt          -- has a vcpkg port which needs to be thoroughly refactored (conflicts with libsquish and others)
+#   * pvrtc         -- https://bitbucket.org/jthlim/pvrtccompressor/
```
Can these comments be removed here?

NancyLi1013 commented on ports/bx/portfile.cmake:
```diff
@@ -0,0 +1,31 @@
+    SOURCE_PATH ${SOURCE_PATH}
+    PREFER_NINJA
+    OPTIONS_RELEASE -DBX_INSTALL_TOOLING=1
+    OPTIONS_DEBUG -DBX_CONFIG_DEBUG=1 -DBX_DISABLE_HEADER_INSTALL=ON
```
Could you please update this as the follows?
```
OPTIONS_RELEASE 
    -DBX_INSTALL_TOOLING=1
OPTIONS_DEBUG 
    -DBX_CONFIG_DEBUG=1 
    -DBX_DISABLE_HEADER_INSTALL=ON
```

NancyLi1013 commented:
/azp run

BurningEnlightenment commented on ports/bgfx/portfile.cmake:
```diff
@@ -0,0 +1,82 @@
+        50-cgltf-include.patch
+        60-freetype-include.patch
+        70-glslang-include.patch
+        80-shaderc-vulkan-include.patch
```
FYI I just created an issue asking whether the upstream maintainer would be generally willing to accept the necessary refactorings: bkaradzic/bgfx/issues/2070
However, I'm almost certain that some stuff like the dxsdk patchset will be rejected.

BurningEnlightenment committed:
[bx] Add cmake transitive dependency discovery

BurningEnlightenment committed:
[bimg] Add cmake transitive dependency discovery

BurningEnlightenment committed:
[bgfx] Switch back to vendored ogl/vulkan headers

Upstream bkaradzic/bgfx#2070 recommends using the vendored OpenGL/vulkan
headers, because the respective libraries are only dynamically linked at
runtime, i.e. ODR violations shouldn't be possible considering the
quite minimal C inteface exposed.

- Fix some stylistic nits.
- Correctly include transitive dependencies.

BurningEnlightenment commented on ports/bgfx/portfile.cmake:
```diff
@@ -0,0 +1,82 @@
+
+vcpkg_fail_port_install(ON_TARGET "OSX" "UWP" "ANDROID" ON_LIBRARY_LINKAGE "dynamic")
```
> If this port only supports static build, you can add `vcpkg_check_linkage(ONLY_STATIC_LIBRARY)`.

Well, the upstream project does support a shared build which exposes the c99 API which currently isn't implemented by this cmake project. Should we still simply fallback to static builds?

NancyLi1013 replied:
We can update this port again if it is implemented by this cmake project.
Currently, we can only support static build.

BurningEnlightenment commented:
I'm thinking about generally using the vendored dependencies except for those cases where the dependencies are linked by installed/exported libraries like lodepng, tinyexr, stb, etc. This would allow us to drop the include patches for cgltf, freetype and glslang without any observable side effects. I will however ask again if the other patches can be integrated upstream

NancyLi1013 commented:
/azp run

NancyLi1013 commented:
@BurningEnlightenment 
 Is work still being done for this PR?

NancyLi1013 commented:
/azp run

NancyLi1013 commented:
@BurningEnlightenment 

Thanks for the PR; we're closing this for now since there's been no response. If you'd like to continue working on it, please reopen and ping us!
